File: java/baiduads-sdk-samples/src/main/java/com/baidu/dev2/api/samples/feed/CreateFeedAd.java
package com.baidu.dev2.api.samples.feed;

import com.baidu.dev2.api.sdk.adgroupfeed.api.AdgroupFeedService;
import com.baidu.dev2.api.sdk.adgroupfeed.model.AddAdgroupFeedRequestWrapper;
import com.baidu.dev2.api.sdk.adgroupfeed.model.AddAdgroupFeedRequestWrapperBody;
import com.baidu.dev2.api.sdk.adgroupfeed.model.AddAdgroupFeedResponseWrapper;
import com.baidu.dev2.api.sdk.adgroupfeed.model.AdgroupFeedType;
import com.baidu.dev2.api.sdk.campaignfeed.api.CampaignFeedService;
import com.baidu.dev2.api.sdk.campaignfeed.model.AddCampaignFeedRequestWrapper;
import com.baidu.dev2.api.sdk.campaignfeed.model.AddCampaignFeedRequestWrapperBody;
import com.baidu.dev2.api.sdk.campaignfeed.model.AddCampaignFeedResponseWrapper;
import com.baidu.dev2.api.sdk.campaignfeed.model.CampaignFeedType;
import com.baidu.dev2.api.sdk.campaignfeed.model.ScheduleType;
import com.baidu.dev2.api.sdk.common.ApiRequestHeader;
import com.baidu.dev2.api.sdk.creativefeed.api.CreativeFeedService;
import com.baidu.dev2.api.sdk.creativefeed.model.AddCreativeFeedRequestWrapper;
import com.baidu.dev2.api.sdk.creativefeed.model.AddCreativeFeedRequestWrapperBody;
import com.baidu.dev2.api.sdk.creativefeed.model.AddCreativeFeedResponseWrapper;
import com.baidu.dev2.api.sdk.creativefeed.model.CreativeFeedType;
import com.baidu.dev2.api.sdk.invoke.ApiException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 创建信息流广告示例
 */
public class CreateFeedAd {
    private static final Logger LOGGER = LoggerFactory.getLogger(CreateFeedAd.class);

    private static ApiRequestHeader header;

    static {
        header = new ApiRequestHeader();
        header.setPassword("xxxxx");
        header.setUsername("xxxxx");
        header.setToken("xxxxx");
    }

    /**
     * 新建计划
     *
     * @return
     * @throws ApiException
     */
    public static List<CampaignFeedType> addCampaignFeed() throws ApiException {
        CampaignFeedService campaignFeedService = new CampaignFeedService();
        AddCampaignFeedRequestWrapperBody addCampaignFeedRequestWrapperBody =
                new AddCampaignFeedRequestWrapperBody();
        AddCampaignFeedRequestWrapper addCampaignFeedRequestWrapper = new AddCampaignFeedRequestWrapper();
        // 设置请求header
        addCampaignFeedRequestWrapper.setHeader(header);
        // 构造请求参数
        ArrayList<CampaignFeedType> campaignFeedTypes = new ArrayList<>();
        CampaignFeedType campaignFeedType = new CampaignFeedType();
        campaignFeedType.setCampaignFeedName("xx计划名称");
        // 设置推广对象 1 - 网站链接 2 - 应用下载（IOS）3 - 应用下载（Android）7 - 电商店铺
        campaignFeedType.setSubject(1);
        // 设置计划推广开始结束时段 开始时间不能早于当天日期 结束时间不能早于开始日期
        campaignFeedType.setStarttime("2022-02-23");
        campaignFeedType.setEndtime("2022-09-12");
        // 设置计划推广暂停时段
        ArrayList<ScheduleType> scheduleTypes = new ArrayList<>();
        ScheduleType scheduleType = new ScheduleType();
        scheduleType.setWeekDay(1);
        scheduleType.setStartHour(1L);
        scheduleType.setEndHour(2L);
        scheduleTypes.add(scheduleType);
        campaignFeedType.setSchedule(scheduleTypes);
        campaignFeedTypes.add(campaignFeedType);
        addCampaignFeedRequestWrapperBody.setCampaignFeedTypes(campaignFeedTypes);
        // 设置请求body
        addCampaignFeedRequestWrapper.setBody(addCampaignFeedRequestWrapperBody);
        // 调用API
        AddCampaignFeedResponseWrapper addCampaignFeedResponseWrapper =
                campaignFeedService.addCampaignFeed(addCampaignFeedRequestWrapper);
        // 判断请求是否成功
        if (addCampaignFeedResponseWrapper.getHeader().getStatus() != 0) {
            LOGGER.error("addCampaignFeed fail. errors:" + addCampaignFeedResponseWrapper.getHeader());
            throw new RuntimeException("addCampaignFeed fail.");
        }
        return addCampaignFeedResponseWrapper.getBody().getData();
    }

    /**
     * 在新建计划下新建单元
     *
     * @param campaignFeedId
     * @return
     * @throws ApiException
     */
    public static List<AdgroupFeedType> addAdgroupFeed(Long campaignFeedId) throws ApiException {
        AdgroupFeedService adgroupFeedService = new AdgroupFeedService();
        AddAdgroupFeedRequestWrapperBody addAdgroupFeedRequestWrapperBody =
                new AddAdgroupFeedRequestWrapperBody();
        AddAdgroupFeedRequestWrapper addAdgroupFeedRequestWrapper = new AddAdgroupFeedRequestWrapper();
        // 设置请求header
        addAdgroupFeedRequestWrapper.setHeader(header);
        // 构造请求参数
        ArrayList<AdgroupFeedType> adgroupFeedTypes = new ArrayList<>();
        AdgroupFeedType adgroupFeedType = new AdgroupFeedType();
        adgroupFeedType.setCampaignFeedId(campaignFeedId);
        adgroupFeedType.setAdgroupFeedName("xx单元名称");
        adgroupFeedType.setPause(false);
        // 定向设置
        Map audience = new HashMap();
        audience.put("region", 9999999);
        audience.put("age", 0);
        audience.put("sex", 0);
        audience.put("lifeStage", 0);
        audience.put("education", 0);
        audience.put("newInterests", 0);
        audience.put("keywords", "");
        adgroupFeedType.setAudience(audience);
        adgroupFeedType.setBid(66.66);
        ArrayList<Integer> productTypes = new ArrayList<>();
        productTypes.add(1);
        productTypes.add(2);
        adgroupFeedType.setProducttypes(productTypes);
        ArrayList<Integer> fTypes = new ArrayList<>();
        fTypes.add(1);
        adgroupFeedType.setFtypes(fTypes);
        adgroupFeedTypes.add(adgroupFeedType);
        // 设置优化目标和付费模式
        adgroupFeedType.setBidtype(1);
        addAdgroupFeedRequestWrapperBody.setAdgroupFeedTypes(adgroupFeedTypes);
        // 设置请求body
        addAdgroupFeedRequestWrapper.setBody(addAdgroupFeedRequestWrapperBody);
        // 调用API
        AddAdgroupFeedResponseWrapper addAdgroupFeedResponseWrapper =
                adgroupFeedService.addAdgroupFeed(addAdgroupFeedRequestWrapper);
        // 判断请求是否成功
        if (addAdgroupFeedResponseWrapper.getHeader().getStatus() != 0) {
            LOGGER.error("addCampaignFeed fail. errors:" + addAdgroupFeedResponseWrapper.getHeader());
            throw new RuntimeException("addCampaignFeed fail.");
        }
        return addAdgroupFeedResponseWrapper.getBody().getData();
    }

    /**
     * 在新建单元下新建创意
     *
     * @param adgroupFeedId
     * @return
     * @throws ApiException
     */
    public static List<CreativeFeedType> addCreativeFeed(Long adgroupFeedId) throws ApiException {
        CreativeFeedService creativeFeedService = new CreativeFeedService();
        AddCreativeFeedRequestWrapperBody addCreativeFeedRequestWrapperBody =
                new AddCreativeFeedRequestWrapperBody();
        AddCreativeFeedRequestWrapper addCreativeFeedRequestWrapper = new AddCreativeFeedRequestWrapper();
        // 设置请求header
        addCreativeFeedRequestWrapper.setHeader(header);
        // 构造请求参数
        ArrayList<CreativeFeedType> creativeFeedTypes = new ArrayList<>();
        CreativeFeedType creativeFeedType = new CreativeFeedType();
        creativeFeedType.setAdgroupFeedId(adgroupFeedId);
        creativeFeedType.setCreativeFeedName("xx创意名称");
        // 设置创意样式 取值请参考信息流创意样式列表
        creativeFeedType.setMaterialstyle(101);
        creativeFeedType.setPause(false);
        // 设置物料内容 参考信息流创意物料说明和信息流创意样式列表
        creativeFeedType.setMaterial("请传json");
        creativeFeedType.setIdeaType(0);
        creativeFeedTypes.add(creativeFeedType);
        addCreativeFeedRequestWrapperBody.setCreativeFeedTypes(creativeFeedTypes);
        // 设置请求body
        addCreativeFeedRequestWrapper.setBody(addCreativeFeedRequestWrapperBody);
        // 调用API
        AddCreativeFeedResponseWrapper addCreativeFeedResponseWrapper =
                creativeFeedService.addCreativeFeed(addCreativeFeedRequestWrapper);
        // 判断请求是否成功
        if (addCreativeFeedResponseWrapper.getHeader().getStatus() != 0) {
            LOGGER.error("addCreative fail. errors:" + addCreativeFeedResponseWrapper.getHeader());
            throw new RuntimeException("addCreative fail.");
        }
        return addCreativeFeedResponseWrapper.getBody().getData();
    }

    public static void main(String[] args) {
        try {
            // 第一步：创建计划
            List<CampaignFeedType> campaignFeedTypes = addCampaignFeed();
            LOGGER.info("campaignFeedTypes:{}", campaignFeedTypes);
            // 获取计划id
            Long campaignFeedId = campaignFeedTypes.get(0).getCampaignFeedId();

            // 第二步：在新建计划下新建单元
            List<AdgroupFeedType> adgroupFeedTypes = addAdgroupFeed(campaignFeedId);
            LOGGER.info("adgroupFeedTypes:{}", adgroupFeedTypes);
            // 获取单元id
            Long adgroupFeedId = adgroupFeedTypes.get(0).getAdgroupFeedId();

            // 第三步：在新建单元下新建创意
            List<CreativeFeedType> creativeFeedTypes = addCreativeFeed(adgroupFeedId);
            LOGGER.info("creativeFeedTypes:{}", creativeFeedTypes);
        } catch (Exception e) {
            LOGGER.error("CreateFeedAd fail.", e);
        }
    }
}

File: python/baiduads-sdk-samples/baiduads/samples/feed/create_feed_ad.py
"""
创建信息流广告示例
"""
from baiduads.adgroupfeed.api import AdgroupFeedService
from baiduads.adgroupfeed.model import AddAdgroupFeedRequestWrapper, AdgroupFeedType, AddAdgroupFeedRequestWrapperBody
from baiduads.campaignfeed.api import CampaignFeedService
from baiduads.campaignfeed.model import AddCampaignFeedRequestWrapper, CampaignFeedType, \
    AddCampaignFeedRequestWrapperBody
from baiduads.common.model import ApiRequestHeader
import logging

from baiduads.creativefeed.api import CreativeFeedService
from baiduads.creativefeed.model import AddCreativeFeedRequestWrapper, AddCreativeFeedRequestWrapperBody, \
    CreativeFeedType

logger = logging.getLogger("CreateFeedAd")
logger.setLevel(logging.INFO)
# set STDOUT
logger.addHandler(logging.StreamHandler())


class CreateFeedAd:
    header = ApiRequestHeader(
        token="xxxxx",
        username="xxxxx",
        password="xxxxx",
        _spec_property_naming=True
    )

    @classmethod
    def add_campaign_feed(cls):
        campaign_feed_service = CampaignFeedService()
        # 构造参数
        campaign_feed_type = CampaignFeedType(
            campaignFeedName="推广计划XX",
            subject=1,
            campaignType=1,
            bgtctltype=0,
            pause=False,
            bstype=1,
            eshopType="1",
            _spec_property_naming=True
        )
        body = AddCampaignFeedRequestWrapperBody(
            campaignFeedTypes=[campaign_feed_type],
            _spec_property_naming=True
        )
        add_campaign_feed_request_wrapper = AddCampaignFeedRequestWrapper(
            header=cls.header,  # 设置请求header
            body=body,  # 设置请求body
            _spec_property_naming=True
        )
        # 调用API
        add_campaign_feed_response_wrapper = campaign_feed_service.add_campaign_feed(add_campaign_feed_request_wrapper)
        # 判断请求是否成功
        if add_campaign_feed_response_wrapper.header.status != 0:
            logger.error("add_campaign_feed fail. errors:%s", add_campaign_feed_response_wrapper.header)
            raise RuntimeError("add_campaign_feed fail.")
        return add_campaign_feed_response_wrapper.body.data

    @classmethod
    def add_adgroup_feed(cls, campaign_feed_id):
        adgroup_feed_service = AdgroupFeedService()
        # 构造参数
        adgroup_feed_type = AdgroupFeedType(
            campaignFeedId=campaign_feed_id,
            adgroupFeedName="推广单元XX",
            pause=False,
            bid=1.100000023841858,
            producttypes=[1],
            ftypes=[1],
            bidtype=1,
            _spec_property_naming=True
        )
        body = AddAdgroupFeedRequestWrapperBody(
            adgroupFeedTypes=[adgroup_feed_type],
            _spec_property_naming=True
        )
        add_adgroup_feed_request_wrapper = AddAdgroupFeedRequestWrapper(
            header=cls.header,  # 设置请求header
            body=body,  # 设置请求body
            _spec_property_naming=True
        )
        # 调用API
        add_adgroup_feed_response_wrapper = adgroup_feed_service.add_adgroup_feed(add_adgroup_feed_request_wrapper)
        # 判断请求是否成功
        if add_adgroup_feed_response_wrapper.header.status != 0:
            logger.error("add_adgroup_feed fail. errors:%s", add_adgroup_feed_response_wrapper.header)
            raise RuntimeError("add_adgroup_feed fail.")
        return add_adgroup_feed_response_wrapper.body.data

    @classmethod
    def add_creative_feed(cls, adgroup_feed_id):
        creative_feed_service = CreativeFeedService()
        # 构造参数
        creative_feed_type = CreativeFeedType(
            adgroupFeedId=adgroup_feed_id,
            materialstyle=101,
            creativeFeedName="创意XX",
            pause=True,
            material="{\"title\":\"测试\",\"brand\":\"11个字符111\",\"us"
                     "erPortrait\":\"https://fc-feed.cdn.bcebos.com/0/p"
                     "ic/b70038ce4be0de17b4f051fda732da8b.jpg\",\"pcUrl"
                     "\":\"http://www.baidu.com\",\"url\":\"http://www"
                     ".baidu.com\",\"pictures\":[{\"image\":\"https://"
                     "aod-image.baidu.com/5/pic/d404a1e0252d81958378253"
                     "ae54d45a1.jpg@s_2,w_370,h_245\"}],\"pictureWidth\""
                     ":\"370\",\"pictureHeight\":\"245\",\"subtitle\""
                     ":\"使用信息流推\",\"monitorUrl\":\"http://www.baid"
                     "u.com/{{CALLBACK_URL}}\",\"exposureUrl\":\"http:/"
                     "/www.baidu.com\",\"naUrl\":\"topen://www.baidu.co"
                     "m&{{DEEPLINK_BACKURL}}\",\"ulkUrl\":\"https://ww"
                     "w.baidu.com\",\"ulkScheme\":\"sing://fdfdfdfxxxx\"}",
            ideaType=0,
            _spec_property_naming=True
        )
        body = AddCreativeFeedRequestWrapperBody(
            creativeFeedTypes=[creative_feed_type],
            _spec_property_naming=True
        )

        add_creative_feed_request_wrapper = AddCreativeFeedRequestWrapper(
            header=cls.header,  # 设置请求header
            body=body,  # 设置请求body
            _spec_property_naming=True
        )
        # 调用API
        add_creative_feed_response_wrapper = creative_feed_service.add_creative_feed(add_creative_feed_request_wrapper)
        # 判断请求是否成功
        if add_creative_feed_response_wrapper.header.status != 0:
            logger.error("add_creative_feed fail. errors:%s", add_creative_feed_response_wrapper.header)
            raise RuntimeError("add_creative_feed fail.")
        return add_creative_feed_response_wrapper.body.data


def main():
    # 第一步：创建计划
    campaign_feed_types = CreateFeedAd.add_campaign_feed()
    logger.info("campaign_feed_types:%s", campaign_feed_types)
    # 获取 campaign_feed_id
    campaign_feed_id = campaign_feed_types[0].campaign_feed_id

    # 第二步：在新建计划下新建单元
    adgroup_feed_types = CreateFeedAd.add_adgroup_feed(campaign_feed_id)
    logger.info("adgroup_feed_types:%s", adgroup_feed_types)
    # 获取 adgroup_feed_id
    adgroup_feed_id = adgroup_feed_types[0].adgroup_feed_id

    # 第三步：在新建单元下新建创意
    creative_feed_types = CreateFeedAd.add_creative_feed(adgroup_feed_id)
    logger.info("creative_feed_types:%s", creative_feed_types)


if __name__ == '__main__':
    main()
